### PR TITLE
Ht 114 clip grid to aoi

### PR DESCRIPTION
**Pull request by feenster**
- Created a new api for intersecting task grid with aoi
- added a trim step t project creation wizard

### CODE REVIEW

File: server/api/grid_apis.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+                                              $ref: "#/definitions/GeoJsonFeature"
+        responses:
+            200:
+                description: Intersecting tasks found successfully
+            500:
+                description: Internal Server Error
+        """
+        try:
+            json = request.get_json()
+            grid_dto = GridDTO(json)
```
Comment (hunt3ri) [inline server/api/grid_apis.py]: I would split getting the Grid DTO and calling the service into 2 steps.  Then u can validate the DTO

File: server/api/grid_apis.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+            grid = GridService.find_intersecting_tiles_in_grid(grid_dto.grid, grid_dto.area_of_interest)
```
Comment (hunt3ri) [inline server/api/grid_apis.py]: Better to pass the grid dto as a whole object, as means u can extend the object without having to rework the method API

File: server/services/grid_service.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+    @staticmethod
+    def geometries_intersect(tile: shape, aoi: shape) -> bool:
+        """
+        Takes two shapes and returns True if the intersect, False if not
+        :param tile: shapely.geometry.shape
+        :param aoi: shapely.geometry.shape
+        :return: bool
+        """
+        p = prep(aoi)
```
Comment (hunt3ri) [inline server/services/grid_service.py]: Single letter variable 2 demerit points 👎 😉 
Rename to something more descriptive

File: tests/server/unit/services/test_grid_service.py [hunk 4]
```diff
@@ -0,0 +1,865 @@
+import unittest
+from server.services.grid_service import GridService
+from shapely.geometry import shape
+
+
+class TestGridService(unittest.TestCase):
+
+    def test_geometries_intersect(self):
+        # arrange
+        task = {
```
Comment (hunt3ri) [inline tests/server/unit/services/test_grid_service.py]: Suggest loading JSON from a file, if poss to make code easier to read.

File: server/api/grid_apis.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+            200:
+                description: Intersecting tasks found successfully
+            400:
+                description: Client Error - Invalid Request
+            500:
+                description: Internal Server Error
+        """
+        try:
+            json = request.get_json()
+            grid_dto = GridDTO(json)
```
Comment (hunt3ri) [inline server/api/grid_apis.py]: I'd do this in one line:

grid_dto = GridDTO(request.get_json())

File: server/models/dtos/grid_dto.py [hunk 6]
```diff
@@ -0,0 +1,8 @@
+from schematics.types import BaseType, BooleanType
+from schematics import Model
+
```
Comment (hunt3ri) [inline server/models/dtos/grid_dto.py]: 2 blanks lines for PEP8

File: server/services/grid_service.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
+import geojson
+from shapely.geometry import MultiPolygon, mapping
+import shapely.geometry
+from geoalchemy2 import shape
+from server.models.dtos.grid_dto import GridDTO
+from server import db
+from server.models.postgis.utils import ST_Transform
+
```
Comment (hunt3ri) [inline server/services/grid_service.py]: 2 blanks lines for PEP8

File: server/services/grid_service.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+            # created a shapely shape for the tile
+            tile = shapely.geometry.shape(feature['geometry'])
+            if aoi_polygon.contains(tile):
+                # tile is completely within aoi, use as is
+                intersecting_features.append(feature)
+            else:
+                intersection = aoi_polygon.intersection(tile)
+                if not intersection.is_empty:
+                    # tile is partailly intersecting the aoi
+                    if clip_to_aoi:
```
Comment (hunt3ri) [inline server/services/grid_service.py]: Might want to think about extracting a _clip_aoi method here, which would reduce nesting and clarify code a bit?

File: client/app/services/project.service.js [hunk 9]
```diff
@@ -606,5 +604,34 @@
                 return $q.reject("error");
             });
         }
+
```
Comment (LindaAlblas) [inline client/app/services/project.service.js]: please add function docs

File: client/app/admin/create-project/create-project.controller.js [hunk 10]
```diff
@@ -186,29 +187,48 @@
...
+
+            var taskGrid = projectService.getTaskGrid()
+            var trimTaskGridPromise = projectService.trimTaskGrid(vm.clipTasksToAoi)
+            trimTaskGridPromise.then(function (data) {
+                projectService.removeTaskGrid();
+                var tasksGeoJson = geospatialService.getFeaturesFromGeoJSON(data, 'EPSG:3857')
+                projectService.setTaskGrid(tasksGeoJson);
+                projectService.addTaskGridToMap();
+            }, function () {
+                console.log('error')
```
Comment (LindaAlblas) [inline client/app/admin/create-project/create-project.controller.js]: remove console.log

File: client/app/admin/create-project/create-project.controller.js [hunk 11]
```diff
@@ -186,29 +187,48 @@
...
+            })
+
```
Comment (LindaAlblas) [inline client/app/admin/create-project/create-project.controller.js]: remove blank line

File: client/app/admin/create-project/create-project.controller.js [hunk 12]
```diff
@@ -385,13 +405,19 @@
...
-            if (vm.drawAndSelectPoint){
+            if (vm.drawAndSelectPoint) {
                 vm.drawAndSelectPoint.setActive(boolean);
             }
         }
+
+        vm.toggleClipTasksToAoi = function(){
+            vm.clipTasksToAoi = !vm.clipTasksToAoi;
+        }
+
```
Comment (LindaAlblas) [inline client/app/admin/create-project/create-project.controller.js]: remove blank lines

**Comment by LindaAlblas** [comment]
@feenster , could you fix the indicating the steps please?

File: server/api/grid_apis.py (hunk 5 above)
Comment (feenster) [inline server/api/grid_apis.py]: done

File: server/models/dtos/grid_dto.py (hunk 6 above)
Comment (feenster) [inline server/models/dtos/grid_dto.py]: done

File: server/services/grid_service.py (hunk 7 above)
Comment (feenster) [inline server/services/grid_service.py]: done

File: server/services/grid_service.py (hunk 8 above)
Comment (feenster) [inline server/services/grid_service.py]: done

File: client/app/services/project.service.js (hunk 9 above)
Comment (feenster) [inline client/app/services/project.service.js]: done

File: client/app/admin/create-project/create-project.controller.js (hunk 10 above)
Comment (feenster) [inline client/app/admin/create-project/create-project.controller.js]: done

File: client/app/admin/create-project/create-project.controller.js (hunk 11 above)
Comment (feenster) [inline client/app/admin/create-project/create-project.controller.js]: done

File: client/app/admin/create-project/create-project.controller.js (hunk 12 above)
Comment (feenster) [inline client/app/admin/create-project/create-project.controller.js]: done

**Pull request merged**